### DISCUSSION TITLE
Is there a tutorial for beginners

### DISCUSSION THREAD

**Discussion by dzwduan** [sparcians/map#438] [General]
Just  like learning GEM5

**Comment by ghost** [comment]
Are you looking for tutorials on _creating_ a simulator using the Sparta FW or tutorials on _using_ a simulator built with Sparta?

For building a simulator, I've been wanting to create a tutorial in the [wiki](https://github.com/sparcians/map/wiki) section of this repo, but time is a limited resource 😆.  For folks that posted that question to me before, I've asked for guidance on **how** they would like that tutorial to flow.  So far, no response.  😉 

There are a couple of high-level facets to the framework:

1. [Command line](https://sparcians.github.io/map/common_cmdline.html) and [Simulation](https://sparcians.github.io/map/classsparta_1_1app_1_1Simulation.html) interfaces
2. Sparta simulation hierarchy classes ([RootTreeNode](https://sparcians.github.io/map/classsparta_1_1RootTreeNode.html) and [TreeNode](https://sparcians.github.io/map/classsparta_1_1TreeNode.html)) 
3. Sparta [Units](https://sparcians.github.io/map/classsparta_1_1Unit.html) and [Resources](https://sparcians.github.io/map/classsparta_1_1Resource.html)
4. [Ports](https://sparcians.github.io/map/classsparta_1_1Port.html), [Events](https://sparcians.github.io/map/classsparta_1_1Event.html)
5. [Parameters](https://sparcians.github.io/map/classsparta_1_1Parameter.html), [Statistics/Counters](https://sparcians.github.io/map/classsparta_1_1Counter.html)
6. etc...

I can formulate a tutorial that illustrates how these components work together and how timing is modeled, but I need an interesting "machine" to build for that tutorial.  Got any ideas?

In the meantime, I've relied on examples to help folks out.  Specifically, the [Skeleton Pipeline](https://sparcians.github.io/map/skeleton_example.html) and _very_ simple [Core Model](https://sparcians.github.io/map/core_example.html).  These are well documented and illustrate basic functionality of the components listed above.

**Reply by dzwduan** [reply]
I have learned about sparta overview in sparta, i think the pictures in slides is helpful. And i think how to guide beginners to create a five-stage pipeline from scratch is more helpful, because the difficulty of superscalar processors is still a problem. And we need a part that can guide the modification of the underlying implementation, just like gem5 guides simobject.

**Comment by ghost** [comment]
You might also want to check out the RISC-V performance model that's under development over at RISV International: https://github.com/riscv-software-src/riscv-perf-model
